### ISSUE TITLE
Italics aren't displaying properly

### ISSUE THREAD

**Issue by ivywong** [neovim/neovim#3461]
I recently got italics to work with terminal vim, and while trying out neovim, I noticed that none of the italics are showing up (nvim on the left, vim on the right):

![italicscomparison](https://cloud.githubusercontent.com/assets/2830112/10534026/6dfe68fe-739f-11e5-902f-188774798a58.png)

(This occurs with or without tmux.) I don't think it's a problem with my terminal or vimrc, because italics show up fine in regular vim. Does it have to do with neovim ignoring terminal codes? If so, how can I get italics to work?

Note: this is with the latest version of neovim (built Oct 13 2015), gnome-terminal 3.14.2, and Ubuntu 15.04.

**Comment by justinmk** [comment]
what version of VTE is gnome-terminal using? (try looking for libvte in your package manager)

**Comment by ivywong** [comment]
It's either 0.38.3 or 0.36.3--both are installed, but I'm not sure which
one gnome-terminal is using.

**Comment by sittim** [comment]
I get the same thing with `screen-256color`.  Italics work with `xterm-256color` and  `rxvt-unicode-256color`.  To fix in tmux, add the following line to .tmux.conf: `set -g default-terminal "rxvt-unicode-256color"` see below:

![selection_044](https://cloud.githubusercontent.com/assets/8825037/10545299/01320468-73f6-11e5-9313-ca67be7c6134.png)

**Comment by sittim** [comment]
@ivywong

Note that before the changes take effect, you have to end all tmux sessions and restart all of the terminals.

**Comment by justinmk** [comment]
@ivywong what is `echo $TERM`?

@sittim Thanks for that info.

**Comment by ivywong** [comment]
@justinmk @sittim was right. I had $TERM set to `screen-256color`. Changing it to `xterm-256color` made everything work properly. Thanks!

**Comment by tarruda** [comment]
@ivywong I do not recommend using TERM=xterm-256color in tmux because some codes may be incompatible, [here's](https://github.com/neovim/neovim/issues/3462#issuecomment-148729580) a better approach if you need italics.

**Comment by ivywong** [comment]
@tarruda Hmm... I haven't noticed any strange characters yet. I'll try that solution if I do. Thanks for the heads up.

**Comment by sittim** [comment]
I did try that solution, it did not work for me.

**Comment by tarruda** [comment]
@sittim any chance you tried that in gnome terminal? As I explained in my comment, that one was for rxvt-unicode. Gnome terminal probably has different codes for italics

**Comment by sittim** [comment]
@tarruda 

I am probably not knowledgeable enough to know how to not use gnome terminal.  I tried running on ROXTerm,

![selection_050](https://cloud.githubusercontent.com/assets/8825037/10551473/1c93a296-741b-11e5-8c80-c53794d6d33f.png)

but I do get the same effect.  Note:  I do not force a value on to `TERM`, unless it is `xterm`, then I change it to `xterm-256color`.  So to start NeoVIM:

1) lauch roxterm or tilda, running `echo $TERM`, I get `xterm256color`
2) lauch tmux, running `echo $TERM`, I get `rxvt-unicode-256color`

This works OK, except once in a while I get these funny chars.  I never used to get these chars in VIM.  I think that I do not have a good understanding of what is a terminal.  Its confusing, so I can run screen in gnome terminal, which is the multiplexer for the tmux, but then when the tmux is not running, then things are processed through gnome terminal?  So to get a clear picture, I would have to do some digging.

**Comment by EricR86** [comment]
I cannot get this to work with TERM=xterm-256color but I can with TERM=rxvt or any rxvt variant. I use the exact same italic settings as the one @ivywong used in the .vimrc.

I can consistently get italics to work in vim with the same terminal and font and not in nvim.

**Comment by EricR86** [comment]
After looking into this more carefully I've discovered this is more of a distribution / terminal problem.  I can set my $TERM to "gnome-256color" and it solves my issue (I think). The issue is that gnome-terminal reports and sets its $TERM to "xterm-256color". This is the solution gnome-terminal settled on last year after years of discussion. See:

https://bugzilla.gnome.org/show_bug.cgi?id=115750

Neovim reads these terminfo settings "correctly" from my distribution using $TERM and unibilium and tries using them. However in my distribution (CentOS 7) the "xterm-256color" terminfo file doesn't contain entries for italics and some other settings. Presumably regular Vim just guesses correctly somehow that my system does.

Fortunately there is a terminfo database entry in my distribution for "gnome-256color" (in /usr/share/terminfo) which looks like it more accurately represents the capabilities of the gnome-terminal but it's hard to be certain. There are a ton of entries to check or cross reference with "xterm-256color". If you're having the same problem you can run "infocmp" on both "xterm-256color" and "gnome-256color" to see all the differences yourself.

AFAIK there is no way to lauch gnome-terminal and have it report it's $TERM anything other than "xterm-256color". This is a pain if you SSH into your machine and your .bashrc sets $TERM (to an incorrect value). The only reasonable workaround I've found is using an alias:

alias gnvim="TERM=gnome-256color nvim"

I cannot find anywhere on the internet for any accurate terminfo file for gnome-terminal. In my case  "gnome-256color" is simply my best guess.
I'm really not certain what a better solution is than this. I certainly don't want to use this alias if I ssh in.

**Comment by EricR86** [comment]
It's also worth noting that $COLORTERM is no longer set by gnome-terminal which has been used in the past to detect if gnome-terminal is being run.

**Comment by mong8se** [comment]
I have noticed that if I have set NVIM_TUI_ENABLE_TRUE_COLOR then italics don't work.

If I unset NVIM_TUI_ENABLE_TRUE_COLOR then italics do work.

NVIM 0.1.4 on El Capitan

**Comment by skattyadz** [comment]
Can confirm @mong8se's findings

**Comment by ZyX-I** [comment]
You have outdated Neovim: NVIM_TUI_ENABLE_TRUE_COLOR was removed in master, after my guicolors patch was merged in Vim (and thus Vim received a setting which could be used in Neovim in place of NVIM_TUI_ENABLE_TRUE_COLOR). Though I am not sure whether anybody fixed this.

**Comment by skattyadz** [comment]
Thanks for the info.
I've updated to HEAD and can confirm that having `set termguicolors` also disables italic text

**Comment by wsdjeg** [comment]
@ivywong can you add `let  g:gruvbox_italic=1` to your vimrc? I also using gruvbox, but it works well for me with neovim.

**Comment by wsdjeg** [comment]
see https://github.com/morhetz/gruvbox/blob/d7f7f1951d0716957f3e32549a247e92d25b1449/colors/gruvbox.vim#L31-L37

**Comment by vbauerster** [comment]
As of tmux 2.1, if default-terminal is set to "screen" or matches "screen-*",
tmux will behave like screen and italics will be disabled.
https://github.com/tmux/tmux/blob/2.1/FAQ#L355-L383

**Comment by pearofducks** [comment]
Can confirm this is also happening in MacOS, italics are working in Vim and not Neovim. I have italics  enabled through:

```
let &t_ZH="\e[3m"
let &t_ZR="\e[23m"
```

**Comment by lunks** [comment]
If I change iTerm2 to `rxvt`, it works, but no on `xterm-256color` terminal emulation.

**Comment by mhinz** [comment]
There's no `t_*` variables in Neovim anymore. Neovim uses whatever the terminfo says the terminal is offering.

If you have to set `t_ZH` or `t_ZR` for Vim, it only means that your terminfo is missing these entries. You can check for the ritm/sitm entries like this:

```
$ infocmp -L | grep italics_mode
```

As noted above, they should be set to `\e[3m` and `\e[23m`.

I have a short script that fixes backspace and italics for `xterm-256color` and `tmux-256color`: [script](https://github.com/mhinz/dotfiles/blob/master/bin/fix-terminfo).

**Comment by pearofducks** [comment]
@mhinz - confirming your script works

iTerm2 / xterm-256color

**Comment by EricR86** [comment]
Is this issue expected to be resolved by users (or distros) fixing their terminfo settings?

I'm honestly not sure since I don't think it's safe to expect that terminfo settings are not going to be changed or updated by distros and I could reasonably see just adding in "sensible" escape codes (italics and underline) if they don't exist for set of detected terminals regardless of the terminfo they advertise.

**Comment by mhinz** [comment]
@EricR86 

The `t_X` options were removed, because the way Neovim processes keys was completely changed in Neovim (opposed to Vim). This was done for long-term advantages like having more possible key combinations etc.

We want to reintroduce the possibility of customizing entries from within Neovim, but currently you'll have to fix your terminfo database.

And some things Neovim will solve for you automatically, e.g. https://github.com/neovim/neovim/pull/5758

**Comment by justinmk** [comment]
@mhinz Perhaps a healthcheck could be written for this?

> Is this issue expected to be resolved by users (or distros) fixing their terminfo settings?

Is setting a terminfo entry much different than setting a `t_X` option?

**Comment by EricR86** [comment]
@mhinz 
I wasn't suggesting using `t_X` options to solve this issue. I fully agree that they should go.

> And some things Neovim will solve for you automatically, e.g. #5758 

There is a terminfo override in those changes. Why not have an terminfo override for italics? There's already a "[fix_terminfo](https://github.com/neovim/neovim/blob/master/src/nvim/tui/tui.c#L818)" function and this would seem to be a prime candidate for addition.

> Is setting a terminfo entry much different than setting a t_X option?

I think there's a very subtle but important difference between the two that's very relevant here. How many neovim users are going to see that they are missing italics or underlines compared to vim and first think the problem lies in their terminfo file? I'd be more likely to believe that users are going to think there's a problem with the *editor* configuration and not the terminal configuration even though that's technically not true at all.

**Comment by joelostblom** [comment]
If anyone else is trying to get italics working inside tmux and you have python installed via anaconda, it is enough just to change the default terminal line in `~/.tmux.conf` to `set -g default-terminal "tmux-256color"`. The terminfo file will be loaded from `~/anaconda3/terminfo/t/tmux-256-color`. This fixes italics, but does not fix the `^h` issue.

None of the other solutions worked for me, I suspect the anaconda setup just overwrites any modifications done manually with via `tic`.

**Comment by justinmk** [comment]
> This fixes italics, but does not fix the ^h issue.

Ctrl-H is fixed on master.

**Comment by joelostblom** [comment]
Thanks, I noticed that previously. I just wanted to be clear in regards to what was included in the anaconda terminfo file.

**Comment by jdebp** [comment]
Please review this issue to see what problems if any remain when #6816 is in place.

**Comment by justinmk** [comment]
Assuming this is closed by #6816.

**Comment by jsatk** [comment]
As of this writing in the latest iTerm on macOS and the latest nvim in Homebrew if I have `set termguicolors` on in neovim italics doesn't work, they work fine in regular vim. I've tried every single `$TERM` trick in here. All of them. I haven't even started digging into tmux because it doesn't even work outside of tmux.

I had italics working fine before I turned on `set termguicolors` recently.

Does anyone know what's broken in nvim that's causing italics not to work with `set tgc` is on? But works fine in vim?
